text/0000-never-patterns.md

```diff
@@ -0,0 +1,250 @@
+
+Alternatively, we could allow omitting all empty arms; this is what the unstable `exhaustive_patterns` feature does today. A match like
+```rust
+match ... {
+    Ok(_) => ...,
+}
+```
+would implicitly contain a `Err(!)` arm, i.e. would trigger UB if the discriminant is `Err`. This was deemed a footgun and was the original motivation for this feature.
+
+
+# Prior art
+[prior-art]: #prior-art
+
+This proposal is entirely based on @nikomatsakis and @RalfJung's original idea, presented in Niko's [blog post](https://smallcultfollowing.com/babysteps/blog/2018/08/13/never-patterns-exhaustive-matching-and-uninhabited-types-oh-my/).
+
+The only other language I (Nadrieril) am aware of that has the intersection of features that would make this necessary is Zig. They are in the process of clarifying the semantics of their empty types (https://github.com/ziglang/zig/issues/15909); they may or may not end up encountering the same explicitness problem as us.
+
+Ocaml and Adga have respectively [refutation cases](https://ocaml.org/manual/5.2/gadts-tutorial.html#s:gadt-refutation-cases) and [absurd patterns](https://agda.readthedocs.io/en/latest/language/function-definitions.html#absurd-patterns), both of which give a way to say "this pattern is impossible and I shouldn't have to write an arm for it". Their flavor of impossibility is related to types however, not runtime validity.
+
+Something that's suprisingly semantically (and syntactically!) close is Haskell's [bang patterns](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/strict.html#bang-patterns-informal). Haskell is a lazy language and a pattern only evaluates the underlying value if needed to decide which branch to take. A `!x` instead pattern forces the evaluation of the underlying value. Rust's patterns are a bit lazy too: they lazily assert validity of the data they access. In Haskell one must be careful to evaluate the right things in the right order else you may diverge; in unsafe Rust similar care is needed to avoid UB.
+
+Never patterns could then be described as playing the role of both a bang pattern ("please evaluate this") and an absurd pattern ("this is an impossible case") at the same time.
+
+From [the lazy patterns literature](https://dl.acm.org/doi/abs/10.1145/3408989) we can also take the distinction between a "redundant" pattern (can be removed without changing semantics), and an "inaccessible" pattern (will never execute but cannot be removed without changing semantics). E.g., in rust:
+```rust
+let ptr: *const ! = ...;
+match *ptr {
+    // This causes a read of the value, hence UB if reached. Removing it would remove UB. Therefore this is inaccessible and not redundant.
+    _x => {}
+    _ => {}
+}
+```
+Never patterns are our way to notate inaccessible patterns.
+
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Does this cause problems with macros-by-example?
```

Comment:
Yes.  It also causes problems for derive macros.  As [I write](https://github.com/rust-lang/rust/issues/131452#issuecomment-2470433340),

> \[It] sets macro authors up for failure. Every macro-generated match statement needs to have a spurious ! arm, or the macro will fail to compile, sometimes, with an empty enum.

This isn't just a problem for macro_rules macros.  Derive macros can suffer from it too.  (Users of [derive-deftly](https://lib.rs/crates/derive-deftly) or other approaches for making derive macros more easily are quite likely to run into this.)

Reply:
There are two separate issues here: https://github.com/rust-lang/rust/issues/131452 is entirely independent from this RFC, since this RFC is purely about adding a new pattern syntax (and not about changing anything else in the language). The other issue with macros is that a `!` pattern doesn't take an arm; this is the point that's relevant to this RFC.

Reply:
> The other issue with macros is that a ! pattern doesn't take an arm; this is the point that's relevant to this RFC. That one I don't know how bad it is.

I think it would be a breaking change for libraries like [syn](https://crates.io/crates/syn), in the sense that all procedural macro crates would need to update to the latest `syn` with a new AST to handle match patterns without arms - and this may require a major version bump of `syn`...

Given that `syn` is only on major version 2, it looks like such changes aren't very frequent - but maybe this isn't that much of a problem in practice.

Reply:
It is already causing problems like:

```rust
fn h(a: Option<&!>) -> bool {
    matches!(a, Some(!))
    // error: a never pattern is always unreachable
}
```

Reply:
`syn` has to have some way of dealing with the Rust syntax evolving, I don't think its API choices should dictate our grammar.

Reply:
`matches!(a, Some(!))` is kind of silly so arguably we should at the very least lint against it. Why not just write `false` instead?